sw/airborne/modules/ctrl/object_tracking.c

```diff
@@ -0,0 +1,161 @@
+#endif
+
+// Timout in seconds before entering search mode
+#ifndef OBJECT_TRACKING_TIMEOUT
+#define OBJECT_TRACKING_TIMEOUT 3.0f
+#endif
+
+// Deadband (constant heading setpoint)
+#ifndef OBJECT_TRACKING_DEADBAND
+#define OBJECT_TRACKING_DEADBAND RadOfDeg(5)
+#endif
+
+// Turn rate in tracking mode (rad/s)
+#ifndef OBJECT_TRACKING_RATE
+#define OBJECT_TRACKING_RATE RadOfDeg(10)
+#endif
+
+// Turn rate in search mode (rad/s)
+#ifndef OBJECT_TRACKING_SEARCH_RATE
+#define OBJECT_TRACKING_SEARCH_RATE RadOfDeg(20)
+#endif
+
+// Select rate or heading tracking mode
+#ifndef OBJECT_TRACKING_USE_RATE
+#define OBJECT_TRACKING_USE_RATE FALSE
+#endif
+
+// Send debug message
+#ifndef OBJECT_TRACKING_DEBUG
+#define OBJECT_TRACKING_DEBUG TRUE
+#endif
+
+#if OBJECT_TRACKING_DEBUG
+#include "subsystems/datalink/downlink.h"
+#include "pprzlink/messages.h"
+#include "mcu_periph/uart.h"
+#endif
+
+int16_t object_tracking_deadband;
+float object_tracking_rate;
+float object_tracking_search_rate;
+
+float object_lat;
+float object_vert;
+float timeout;
+
+abi_event object_ev;
+
+static const float nav_dt = 1. / NAV_FREQ;
+
+// callback on follow me message
+static void get_object(uint8_t sender_id __attribute__((unused)),
+    uint32_t id __attribute__((unused)),
+    uint8_t frame __attribute__((unused)),
+    float heading, float height,
+    float distance __attribute__((unused)))
+{
+  object_lat = heading;
+  object_vert = height;
+  timeout = 0.f;
+}
+
+void object_tracking_init(void)
+{
+  object_tracking_deadband = OBJECT_TRACKING_DEADBAND;
+  object_tracking_search_rate = OBJECT_TRACKING_SEARCH_RATE;
+  object_tracking_rate = OBJECT_TRACKING_RATE;
+
+  object_lat = 0.f;
+  object_vert = 0.f;
+  timeout = OBJECT_TRACKING_TIMEOUT; // start in search mode
+
+  // Bind to camera message
+  AbiBindMsgFOLLOW_ME(OBJECT_TRACKING_ID, &object_ev, get_object);
+}
+
+void object_tracking_run(void)
+{
+#if OBJECT_TRACKING_USE_RATE
+  float rate = object_tracking_search_rate; // search mode
+  if (timeout < OBJECT_TRACKING_TIMEOUT) {
+    timeout += nav_dt;
+    if (object_lat > object_tracking_deadband) {
+      rate = object_tracking_rate;
+    } else if (object_lat < -object_tracking_deadband) {
+      rate = -object_tracking_rate;
+    } else {
+      rate = 0.f;
+    }
+#if OBJECT_TRACKING_DEBUG
+    float msg[] = {
+      rate,
+      object_lat,
+      object_vert,
+      timeout
+    };
+    DOWNLINK_SEND_PAYLOAD_FLOAT(DefaultChannel, DefaultDevice, 4, msg);
+#endif
+  }
+  nav_heading += ANGLE_BFP_OF_REAL(rate * nav_dt);
+
+#else // use calculated heading
+
+  if (timeout < OBJECT_TRACKING_TIMEOUT) {
+    timeout += nav_dt;
+    // compute expected heading
+    float target_heading = stateGetNedToBodyEulers_f()->psi + object_lat;
+    float diff = target_heading - ANGLE_FLOAT_OF_BFP(nav_heading);
+    FLOAT_ANGLE_NORMALIZE(diff);
+    BoundAbs(diff, object_tracking_rate * nav_dt)
+    nav_heading += ANGLE_BFP_OF_REAL(diff);
```

Comment:
Note that using the difference between the target_heading and the nav_heading rather than the actual vehicle heading will limit the set-point rate not the real rate of the vehicle. I don't think this particularly a problem here, basically reference generator for the set-point. My guess is that was intentional but just wanted to mention it here for future reference.

Reply:
it was intentional, it is the user that should set this rate limit compatible with rotorcraft capabilities